### DISCUSSION TITLE
Faster way to insert into SQL Server

### DISCUSSION THREAD

**Discussion by chivalry123** [sqlalchemy/sqlalchemy#6840] [Usage Questions] [answered]
I'm currently using this method to bulk insert into SQL Server:

    from sqlalchemy import event
    @event.listens_for(engine, "before_cursor_execute")
    def receive_before_cursor_execute(
           conn, cursor, statement, params, context, executemany
            ):
                if executemany:
                    cursor.fast_executemany = True
    
    df1_compute.to_sql('timeseries_1', engine, if_exists="replace", index=True)

This takes 22 seconds to insert 800k lines. But I think the speed is still not fast enough. Is it already the fastest way to bulk insert to SQL Server? Previously, by doing right click and import files with SQL Server Management Studio. It seems I achieve even faster speed

**Comment by zzzeek** [comment, marked as answer]
> This takes 22 seconds to insert 800k lines. But I think the speed is still not fast enough.

Ok what makes you think that?  did you run your query with pyodbc directly?   Write a program that uses [pyodbc directly](https://github.com/mkleehammer/pyodbc/wiki/Getting-started#inserting-data).  Then see how long that takes.   Say it takes 19 seconds , not 22, for 800k lines.  That's still pretty close to just as slow, and if you think that's "not fast enough", [post a bug with pyodbc](https://github.com/mkleehammer/pyodbc/issues), because this has little to do with SQLAlchemy.

OTOH, let's say you get a pyodbc program that inserts 800K rows in 8 seconds, and SQLAlchemy is 22 seconds.  ouch! show us both programs here and we will show you how to get the same form of query that you are using with raw pyodbc as you are getting with SQLAlchemy.

In short, all performance is relative.  Comparing SQLAlchemy to a bulk insert tool in SQL Server management studio is not a reasonable comparison, the bulk insert tooling that comes with databases is by definition dramatically faster than using a client/server library.   you need to compare to raw pyodbc because that's the best we can do.
